src/core/device.c

```diff
@@ -114,12 +156,129 @@ static void device_done(Unit *u) {
+         * In both cases the kernel will automatically fill all other io_cost_qos
+         * parameteters so we do not need to input them */
+        if (io_cost_qos->ctrl == IO_COST_CTRL_AUTO || !enabled) {
+                log_info("IOCostQOSCtrl=auto or not set, setting all other values to kernel default");
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s\n",
+                         major(dev),
+                         minor(dev),
+                         enabled,
+                         io_cost_ctrl_to_string(IO_COST_CTRL_AUTO));
+
+        } else if (io_cost_qos->ctrl == IO_COST_CTRL_USER) {
+                if (io_cost_qos->min > io_cost_qos->max)
+                         return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                "IOCostQOSMin= is greater than IOCostQOSMax=, refusing to apply io.cost.qos settings");
+
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s rpct=%u.%02u rlat=%u wpct=%u.%02u wlat=%u min=%lu.%02lu max=%lu.%02lu\n",
+                         major(dev),
+                         minor(dev),
+                         enabled,
+                         io_cost_ctrl_to_string(io_cost_qos->ctrl),
+                         io_cost_qos->read_latency_percentile / 100,
+                         io_cost_qos->read_latency_percentile % 100,
+                         io_cost_qos->read_latency_threshold,
+                         io_cost_qos->write_latency_percentile / 100,
+                         io_cost_qos->write_latency_percentile % 100,
+                         io_cost_qos->write_latency_threshold,
+                         LOADAVG_INT_SIDE(io_cost_qos->min),
+                         LOADAVG_DECIMAL_SIDE(io_cost_qos->min),
+                         LOADAVG_INT_SIDE(io_cost_qos->max),
+                         LOADAVG_DECIMAL_SIDE(io_cost_qos->max));
+
+        } else
+                return 0;
+
+        r = cg_set_attribute("io", root_slice_unit->cgroup_path, "io.cost.qos", buf);
+        if (r < 0)
+                log_unit_full_errno(root_slice_unit, LOG_LEVEL_CGROUP_WRITE(r), r, "Failed to set '%s' attribute on '%s' to '%.*s': %m",
+                                    strna("io.cost.qos"), empty_to_root(root_slice_unit->cgroup_path), (int) strcspn(buf, NEWLINE), buf);
+
+
+        return 0;
+}
+
+static int root_cgroup_apply_block_device_io_cost_model(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+12+(9+DECIMAL_STR_MAX(uint64_t)+1)*6];
+        IOCostModel *io_cost_model;
+        dev_t dev;
+        int r;
+
+        assert(root_slice_unit);
+        assert(d);
+
+        io_cost_model = &d->io_cost_model;
+
+        /* If ctrl is invalid it means that io_cost_model is not configured
+         * for this device */
+        if (io_cost_model->ctrl == _IO_COST_CTRL_INVALID)
+                return 0;
+
+        r = lookup_block_device(d->devname, &dev);
```

Comment:
the sd_device object can give you the major/minor directly, please use that. i.e. `sd_device_get_devnum()`.

no need to first translate the device node path to major/minor...